### PR TITLE
Simplify public API of the `par_iter` module

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is a first stab at simplifying the public API of the `par_iter` module. Here are the rules I used:

- Expose everything "flat" in the `par_iter` module except for the `internal` traits (`Producer`, etc). 
- Expose only those things that appear in the public interface of the `ParallellIterator` traits.
    - In particular, things like `EnumerateProducer` do not need to be exposed, because the `ExactParallelIterator` trait uses the "double indirection" so that it is never named.
    - To aid in this goal, I adopted the convention of never using `foo::` in public API surface, so that I am forced to `pub use, and **only** using `foo::` inside method bodies, so that I never need to `use`.

I'm pretty happy with how it turned out, but I did run into some questions:

- [x] What should we call the `internal` module? That name seems not great.
- [x] Should the internal APIs remain `#[doc(hidden)]`?
- [x] Do we care that the `new` methods on types like `Enumerator` are publicly surfaced? We could avoid this with more effort (e.g., use a free-function).
- [x] Do we want to expose the "free function" form of things like `find()` (e.g., `find::find()`). I am inclined to say "no" unless there is a demand. Similarly for things like the producer types.
- [x] Naming convention: I used `Combinator` if it is a parallel iterator combinator (e.g., `Enumerate`, `Zip`) and `CollectionIter` if is a `ParallelIterator` implementation for a collection (e.g., `SliceIter`, `RangeIter`). Before we had a mix. This seems to be in the *spirit* of the libstd conventions. However, I am not crazy about the `CollectionIter` convention. I also considered `par_iter::collection::ParIter`, or `CollectionParIter`. Thoughts?

### CODE REVIEW

**Comment by nikomatsakis** [comment]
cc #228 -- this is at least a partial fix, and maybe a complete one.

cc @cuviper

**Comment by cuviper** [comment]
> What should we call the internal module? That name seems not great.

I agree it's not great, but I don't have better ideas. `imp` (implementor)? `low` (as in low-level)?

> Should the internal APIs remain #[doc(hidden)]?

No, we need to show them IMO, but we should make sure they're at the end of the list.

> Do we care that the new methods on types like Enumerator are publicly surfaced? We could avoid this with more effort (e.g., use a free-function).

It's not a *big* deal, but I would prefer a free function.

> Naming convention

I think your choice is fine, and the consistency is good.

File: src/par_iter/mod.rs [hunk 1]
```diff
@@ -14,52 +14,69 @@
...
+mod slice;
+pub use self::slice::{SliceIter, ChunksIter};
+mod slice_mut;
+pub use self::slice_mut::{SliceIterMut, ChunksMutIter};
+mod string;
+pub use self::string::{ParallelString, Pattern, ParChars, ParSplit,
+                       ParSplitTerminator, ParLines,
+                       ParSplitWhitespace};
+mod map;
+pub use self::map::{Map, MapOp, MapFn, MapCloned, MapInspect};
```
Comment (cuviper) [inline src/par_iter/mod.rs]: I don't love having this `Map` hierarchy exposed.  Can we instead have specific `Cloned`, `Inspect`, `Map` (formerly `MapFn`) types, and hide the common `Map` in implementation?

File: src/par_iter/mod.rs [hunk 2]
```diff
@@ -14,52 +14,69 @@
...
-pub mod vec;
-pub mod option;
-pub mod collections;
-pub mod noop;
-pub mod rev;
+mod for_each;
+mod fold;
+pub use self::fold::Fold;
+mod reduce;
+pub use self::reduce::{ReduceOp, SumOp, ProductOp};
```
Comment (cuviper) [inline src/par_iter/mod.rs]: I don't love the `ReduceOp` hierarchy either, but I think we might be able to do away with the whole lot in #252.

File: src/par_iter/mod.rs [hunk 3]
```diff
@@ -14,52 +14,69 @@
...
+mod take;
+pub use self::take::Take;
+mod slice;
+pub use self::slice::{SliceIter, ChunksIter};
+mod slice_mut;
+pub use self::slice_mut::{SliceIterMut, ChunksMutIter};
+mod string;
+pub use self::string::{ParallelString, Pattern, ParChars, ParSplit,
+                       ParSplitTerminator, ParLines,
+                       ParSplitWhitespace};
```
Comment (cuviper) [inline src/par_iter/mod.rs]: Should these get `String`-specific names somehow?  Or maybe this is a large enough chunk of functionality to actually deserve a submodule, just `pub mod string;` and be done.

File: src/par_iter/mod.rs [hunk 4]
```diff
@@ -14,52 +14,69 @@
...
+mod reduce;
+pub use self::reduce::{ReduceOp, SumOp, ProductOp};
+mod skip;
+pub use self::skip::Skip;
+mod take;
+pub use self::take::Take;
+mod slice;
+pub use self::slice::{SliceIter, ChunksIter};
+mod slice_mut;
+pub use self::slice_mut::{SliceIterMut, ChunksMutIter};
```
Comment (cuviper) [inline src/par_iter/mod.rs]: I mentioned I wanted to prototype making the slice extensions to look more like `ParallelString` -- if we do that, I think a `pub mod slice` makes sense too.

**Comment by cuviper** [comment]
While I'm thinking of the module layout, we could flatten this too:

- `rayon`
    - `prelude` should be emphasized
    - `iter`  (`par_` is somewhat redundant)
    - `slice`
    - `string`
    - `internal` (wearing its new bikeshed color)

I think having modules as siblings is cleaner, and also makes a little bit of a test that the necessary functionality is all public, whereas submodules get to cheat a little.

**Comment by cuviper** [comment]
FWIW, I'd be happy with iterating these points over multiple PRs, especially since one massive PR will be disruptive to any other changes we might be considering in the meantime.  So if you want, I say go ahead and merge this as a partial fix for #228, to be continued...

**Comment by nikomatsakis** [comment]
@cuviper ok, I agree, I think I'll apply some of your changes, but land it roughly as-is, and we can then discuss the precise setup on #228.

**Comment by nikomatsakis** [comment]
@cuviper 

ok, so I adjusted the module hierarchy more-or-less as you requested. The one thing I did not do is move or rename `internal` -- I think I prefer it to be caller `rayon::iter::internal` rather than just `rayon::internal`, since its contents are basically specific to the parallel iterator traits, right? That is, I view these as being "part of" the parallel iterator traits.

I also left `MapOp`, `ReduceOp`, and friends as they are.

So the current module hierarchy becomes:

- rayon
    - iter
    - option
    - prelude
    - range
    - slice
    - string
    - vec

**Comment by nikomatsakis** [comment]
@cuviper ok, I think I addressed most everything. I'm going to merge this, presuming that the CI is happy.

Note: I did not leave "deprecated" paths for things that were moved around. That's a pain and I don't know how to do it. =)

**Pull request merged**